File: go_for_data_engineering/05_design_patterns/04_strategy.go
// Strategy: swap algorithm at runtime via interface.
package main

import "fmt"

type Compressor interface {
	Compress(data []byte) []byte
}

type gzipC struct{}

func (gzipC) Compress(d []byte) []byte { return append([]byte("gz:"), d...) }

type zstdC struct{}

func (zstdC) Compress(d []byte) []byte { return append([]byte("zstd:"), d...) }

func Save(data []byte, c Compressor) {
	fmt.Println(string(c.Compress(data)))
}

func main() {
	Save([]byte("hello"), gzipC{})
	Save([]byte("hello"), zstdC{})
}

File: extra_learning/design_patterns/04_strategy.py
"""Strategy: swap algorithm at runtime."""
from __future__ import annotations
from typing import Protocol


class Compressor(Protocol):
    def compress(self, data: bytes) -> bytes: ...


class Gzip:
    def compress(self, data: bytes) -> bytes: return b"gz:" + data


class Zstd:
    def compress(self, data: bytes) -> bytes: return b"zstd:" + data


def save(data: bytes, c: Compressor) -> None:
    print(c.compress(data))


if __name__ == "__main__":
    save(b"hello", Gzip())
    save(b"hello", Zstd())
